### ISSUE TITLE
Migrate existing (JUnit4) tests to JUnit5

### ISSUE THREAD

**Issue by thodorisbais** [unitsofmeasurement/unit-api#154]
As mentioned at #152 , `unit-api` is now JUnit5-ready.
Hence, we can now begin with the migration of currently existing (and written in JUnit4) tests to JUnit5.

Could we assign this to me @keilw ?

**Comment by keilw** [comment]
Yes, I hope the JCP membership question gets answered soon.

**Comment by thodorisbais** [comment]
I was about to say "crossing my fingers", but they're already a bit numb after having them crossed for the last two months 😂 . Anyway, let's hope 2019 will be a better one; we'll definitely chase it as much as we can.

On topic, I think I'm already part of the Observers Team, am I wrong?

**Comment by keilw** [comment]
Observers is mainly for passive following. There was or still is a JCP category but it does not require any membership, joining JCP.org or Oracle.com with an account is usually enough. So that team is a bit of a relic, Contributors match the Contributor membership for Associate JCP Members.

**Comment by keilw** [comment]
@thodorisbais I hope you find a way to fix the code coverage dropping to zero after switching to JUnit5? All quality gates will fail even if the build itself passes. This is not a good way to start the new year, but hope it can be addressed soon, otherwise we'll have to stick with JUnit4.

**Comment by thodorisbais** [comment]
@keilw am I missing something? For me, I see the coverage still ~84% (both for my local and remote repos).
In addition, on a very quick investigation (given the time I had today), I tried adding the `junit-platform-surefire-provider` dependency (as suggested/done in indriya), but nothing changed.

**Comment by thodorisbais** [comment]
Any update on this one @keilw ? As I mentioned, it looks good/same to me
![screenshot 2019-01-05 at 20 30 52](https://user-images.githubusercontent.com/5684688/50728377-f2b7b300-1128-11e9-94d3-5b8de37654f2.png)

**Comment by keilw** [comment]
The JaCoco Plugin and Codacy notice Zero coverage, that's why the code quality also went from A to C: https://app.codacy.com/project/unitsofmeasurement/unit-api/dashboard. It looks better in Indriya which also uses JUnit5, so it should be possible.

**Comment by thodorisbais** [comment]
Can you help me to understand the above ( = how coverage dropped to zero - where is this visible?) ? The only thing that I see in codacy regarding coverage is that it needs to be setup (and I don't have any write permissions to try playing around a bit) : 

![screenshot 2019-01-09 at 23 37 21](https://user-images.githubusercontent.com/5684688/50934194-39c6e080-1468-11e9-8ae3-58b24a3b1e46.png)

**Comment by keilw** [comment]
If you set the `COVEREDRATIO` of the `jacoco-maven-plugin` in the pom.xml to something higher than 0.0 the build fails. It was higher all the time as the version history tells.

**Comment by keilw** [comment]
I sent you an invite to join the uom organization at Codacy. If that works, I can check how to add you to the right teams to allow adjustments.

**Comment by thodorisbais** [comment]
I joined. I can see this for now:

![screenshot 2019-01-10 at 00 15 46](https://user-images.githubusercontent.com/5684688/50935634-f0c55b00-146c-11e9-8a5f-8c0a9bd7862a.png)

**Comment by keilw** [comment]
I'll check when I'm online again. Maybe weekend because I travel and a lot of massive snow in parts of Germany make that unpredictable...

**Comment by thodorisbais** [comment]
Cool! Looking forward to it!

**Comment by thodorisbais** [comment]
I now see what you meant above, Werner. I've reverted my (local) `COVEREDRATIO` back to `0.5` and the tests seem to still run as expected. I don't know how codacy works (I mean, I don't know whether it will still identify errors and we need to take [this extra mile](https://github.com/unitsofmeasurement/indriya/pull/46/files) as you suggested -I guess this is what you meant above by referring to `indriya`), but I hope it runs a build for every new PR; so hopefully, by submitting a PR containing only these changes (reverting `COVEREDRATIO` property back to 0.5) there will be a new build for that PR; then, we take it from there.

**Comment by thodorisbais** [comment]
@keilw as you can see in #160, I've even tried taking the extra mile of the requested addition of [indriya](https://github.com/unitsofmeasurement/indriya/pull/46/files) for this change, but nothing changed.

Are we sure there is something missing from our configuration? Because from my side, all I can see in Codacy is that there is no coverage for any of our projects; that is, I see no coverage graphs/metrics neither for unit-api nor for indriya (but for others as well).

I suspect we are missing some configuration in our Codacy project(s), but on the other hand, if you say it was working before, there is something else going wrong. How could it be that this "something else" applies to all of our projects, though? Are you able to see some coverage graphs/metrics for any of our projects at the moment?

For me, I see nothing both unit-api and indriya, as I fore-mentioned:

![screenshot 2019-01-19 at 12 19 09](https://user-images.githubusercontent.com/5684688/51426000-81760680-1be4-11e9-9c11-c33f5b2e8cdf.png)
![screenshot 2019-01-19 at 12 19 03](https://user-images.githubusercontent.com/5684688/51426001-81760680-1be4-11e9-991a-dbbc7226f729.png)


The only positive that I see for now (in comparison to the beginning of the year) is that at least the project's certification raised from C to B:

![screenshot 2019-01-19 at 12 15 49](https://user-images.githubusercontent.com/5684688/51425977-057bbe80-1be4-11e9-8714-c5231906e4ba.png)

But locally, I see jacoco running properly and the tests being picked up as well; so, a successful build.

Is there something we could do to improve our Codacy configuration? I see that I am missing write permissions to play a bit with it.

**Comment by thodorisbais** [comment]
Moreover, I guess we keep both this and #71 still open for the sake of identifying some coverage metrics on our Codacy project. In case we agree it is a Codacy configuration, though, I would suggest we close both (since both the update to junit5 and migration of existing junit4 tests to junit5 are already done) and create a new issue about fixing our Codacy configuration (coverage-wise).

_Maybe I'm wrong, but it's something to keep in mind, given that we have been keeping two issues open for a month; two issues which (locally) seem to have been resolved. So, either ways, first thing is to identify the root cause of our Codacy's inability to display any coverage metric/graph._

**Comment by thodorisbais** [comment]
I still believe that we shouldn't be discussing the coverage failure of codacy here (; I mean, we should have had a dedicated issue for the codacy failure, instead of commenting here and there on issues related to migration, but given that I haven't seen the past status of codacy and that @keilw claims codacy was working fine before the migration to JUnit5, I will focus on solving our current problem.

That said, I think I just managed to get it working for [my own fork of unit-api](https://app.codacy.com/project/thodorisbais/unit-api/dashboard). The steps I followed can be summarized below:

I followed [this](https://github.com/codacy/codacy-coverage-reporter#setup) guide to set it up (it is a bit outdated, but still relevant enough).

In the end, you just run a command and it automatically sends the coverage report to codacy:
![screenshot 2019-01-20 at 04 39 37](https://user-images.githubusercontent.com/5684688/51434985-8501a000-1c6d-11e9-92d2-0264309db56e.png)
Undoubtedly, this needs to be automated in our build process, but this is just a proof of concept for now(/tonight).

There is a chance I am missing something, so, please correct me if I'm wrong, in case the fore-mentioned procedure is already in place for unit-api.

_By the way, I was playing a bit around at our codacy and given that I wasn't seeing any unit-api listed in our projects (but only being accessible via a url - at least for me), I thought about importing it, but.... :see_no_evil: 
![screenshot 2019-01-20 at 04 45 32](https://user-images.githubusercontent.com/5684688/51435015-491b0a80-1c6e-11e9-8f4f-244a13046fe1.png)

It was of course already there and a replica of it was created (which I of course cannot delete, given my Read rights)_

**Comment by keilw** [comment]
As mentioned, it may be something on Windows only. The CI and the system that screenshot was taken are both Unix, Linux or Mac, so the issue could be Windows-specific.

**Comment by thodorisbais** [comment]
Indeed, now it makes more sense. That means we'd need a Windows machine to make some progress on that.

**Comment by thodorisbais** [comment]
On the other hand, if the CI is Unix, shouldn't it report on a coverage percentage at least? Now, nothing is visible as we fore-mentioned, which gives me the impression that indeed something might not have been properly set up as I fore-mentioned.
Please correct me if I'm wrong.

**Comment by keilw** [comment]
Not sure why, but the unit tests are not executed at all. Maybe something related to the migration, but then why does it work on "NIX" systems like the CI server? I tried to compare it with Indriya where it works locally.

**Comment by keilw** [comment]
They use different coverage tools (with this repo also reporting to Sonar I would not change that now) but both send coverage, the only difference is, Indriya in the log also shows the results of all JUnit test runs in the console. Here it seems in some kind of "silent mode" but since coverage is based on those tests it cannot fail or be ignored.
